hail/python/hail/experimental/vcf_combiner.py

```diff
@@ -323,83 +325,207 @@ def default_exome_intervals(reference_genome='default'):
+        job_str = hl.utils.misc.plural('job', n_jobs)
+        info(f"Starting phase {phase_i}/{n_phases}, merging {len(files_to_merge)} {merge_str} in {n_jobs} {job_str}.")
+
+        if phase_i > 1:
+            intervals = calculate_new_intervals(hl.read_matrix_table(files_to_merge[0]).rows(), config.target_records)
+
+        new_files_to_merge = []
+
+        for job_i, job in enumerate(phase.jobs):
+            job_i += 1  # used for info messages, 1-indexed for readability
+
+            n_merges = len(job.merges)
+            merge_str = hl.utils.misc.plural('file', n_merges)
+            pct_total = 100 * job.input_total_size / total_ops
+            info(f"Starting job {job_i}/{len(phase.jobs)} to create {n_merges} merged {merge_str}, corresponding to ~{pct_total:.1f}% of total I/O.")
+            merge_mts: List[MatrixTable] = []
+            for merge in job.merges:
+                inputs = [files_to_merge[i] for i in merge.inputs]
+
+                if phase_i == 1:
+                    mts = [transform_gvcf(vcf)
+                           for vcf in hl.import_gvcfs(inputs, intervals, array_elements_required=False,
+                                                      _external_header=header,
+                                                      _external_sample_ids=sample_names if header is not None else None)]
+                else:
+                    mts = [hl.read_matrix_table(path, _intervals=intervals) for path in inputs]
+
+                merge_mts.append(combine_gvcfs(mts))
+
+            if phase_i == n_phases: # final merge!
+                assert n_jobs == 1
+                assert len(merge_mts) == 1
+                [final_mt] = merge_mts
+                final_mt.write(out_file, overwrite=overwrite)
+                new_files_to_merge = [out_file]
+                info(f"Finished job {job_i}/{len(phase.jobs)}, {100 * total_work_done / total_ops}% of total I/O finished.")
+                break
+
+            tmp = f'{tmp_path}_phase{phase_i}_job{job_i}/'
+            hl.experimental.write_matrix_tables(merge_mts, tmp, overwrite=True)
+            pad = len(str(len(merge_mts)))
+            new_files_to_merge.extend(tmp + str(n).zfill(pad) + '.mt' for n in range(len(merge_mts)))
+            total_work_done += job.input_total_size
+            info(f"Finished job {job_i}/{len(phase.jobs)}, {100 * total_work_done / total_ops}% of total I/O finished.")
+
+        info(f"Finished phase {phase_i}/{n_phases}.")
+
+        files_to_merge = new_files_to_merge
+
+    assert files_to_merge == [out_file]
+
+    info("Finished!")
+
+def parse_sample_mapping(sample_map_path: str) -> Tuple[List[str], Dict[str, str]]:
+    sample_names: List[str] = list()
+    sample_paths: Dict[str, str] = dict()
+
+    with hl.hadoop_open(sample_map_path) as f:
+        for line in f:
+            [name, path] = line.strip().split('\t')
+            sample_names.append(name)
+            sample_paths[name] = path
```

Comment:
We should make sure there are no duplicate sample names.

Reply:
why? That should work, on?

Reply:
*no?

Reply:
It will work, but I don't think it's what users want. The duplicates will be overwritten with the last path. So you could end up with multiple copies of the same data.

Reply:
I rewrote this to use a list

Reply:
this code won't work for dups, you're right

Reply:
I meant that duplicates should be supported